Generating a generic listing page and blog listing specific page

Pomax opened the pull request:
Closes #3370 
Related PRs/issues #3371 

This does not do any fancy "load initial subset, JS-enabled MORE button to load more results", given that the first use of this model is for blog posts, of which we only have a handful. 

https://github.com/mozilla/foundation.mozilla.org/issues/3443 has been filed for the JS functionality.

### Test

https://foundation-mofostaging-pr-3427.herokuapp.com/en/blog/

alanmoo commented on network-api/networkapi/wagtailpages/models.py:
```diff
@@ -160,6 +161,7 @@ class MiniSiteNameSpace(ModularPage):
         'BanneredCampaignPage',
         'CampaignPage',
         'OpportunityPage',
+        'BlogListingPage',
```
This needs to be a subpage of the Homepage type

Pomax replied:
Just homepage, or also others?

Pomax commented:
@kristinashu this should be ready for design review

Pomax commented on network-api/networkapi/wagtailpages/templates/wagtailpages/listing_page.html:
```diff
@@ -0,0 +1,31 @@
+{% load wagtailcore_tags mini_site_tags %}
+
+
+{% block content %}
```
It appears that the main nav is missing... which is kind of silly.

kristinashu commented:
Thanks Pomax! A few questions and thoughts:
## Tags
For the /blog page, can we show the first tag listed in the posts promote tab for each post?

Comp:
<img width="1155" alt="image" src="https://user-images.githubusercontent.com/1068783/61742925-0cc58580-ad49-11e9-937e-7d9adeacb335.png">

## Filter view
Have you added the filter by tag view yet? I added the tag 'test' to one of the post then tried the url https://foundation-mofostaging-pr-3427.herokuapp.com/en/blog-list/?tag=[test] but it didn't change anything.

Comp: 
<img width="1166" alt="image" src="https://user-images.githubusercontent.com/1068783/61743134-780f5780-ad49-11e9-86db-89bb174515de.png">

## Spacing
Just a bit of finessing is needed for the gaps. I'm reluctant to give exact px spacing but here is a comp with some guidelines. I know this will very slightly because of the preexisting spacing associated with the text so just use this comp as a guideline. Note, we stick to multiples of 8px.

<img width="1235" alt="image" src="https://user-images.githubusercontent.com/1068783/61743628-8ca01f80-ad4a-11e9-8118-f85e2c03d0e8.png">

mmmavis commented:
(Let me know if I should review this sooner. I usually wait for PR to pass design review first.)

Pomax commented:
- tags: can do, but is the first tag always the most appropriate?
- filtering: that's been filed as separate issue, so that while we still have a low number of blog posts we can have a page up that we can refine, rather than wait with landing this until all the parts are done
- spacing wise, I'm relying primarily on the default styling for header/intro, so hopefully the bootstrap padding/margin classes line up with our grid? I'll try to get as close as I can.

Pomax commented:
@mmmavis you can wait until all the design review has been addressed; most of the code won't change, but it's not a huge PR so there's no need to "catch things early on" as it were =)

kristinashu commented:
> tags: can do, but is the first tag always the most appropriate?

I'm open to suggestions but on prod I have gone through each post manually and made sure the first tag is the most relevant and broad.

Pomax commented:
@kristinashu it looks like the spacing around the horizontal rule is fixed in the `intro-and-content-divider` hr class, in a way that adding bootstrap classes for margin/padding can't change (they're not specific enough so the media queries "win" in terms of which rules are applies). I can change this to a different class, but then if we change that rule's styling at some point, this page will not automatically update looks, of course. 

Also for the cards: it's currently using the same spacing/styling that we use for "related cards" on blog pages - do we want to change the spacing specific for this page, for "collections of cards" in general, or stick with the spacing as is to keep it uniform?

kristinashu commented:
> @kristinashu it looks like the spacing around the horizontal rule is fixed in the `intro-and-content-divider` hr class, in a way that adding bootstrap classes for margin/padding can't change (they're not specific enough so the media queries "win" in terms of which rules are applies). I can change this to a different class, but then if we change that rule's styling at some point, this page will not automatically update looks, of course. 

Ok makes sense, I'm ok with keeping the way it is and not doing any overrides.

> Also for the cards: it's currently using the same spacing/styling that we use for "related cards" on blog pages - do we want to change the spacing specific for this page, for "collections of cards" in general, or stick with the spacing as is to keep it uniform?

Updating all "collections of cards" is good!

> (the heading top spacing, at least, was entirely doable with bootstrap instructions =)

Sweet!

Pomax commented:
@alanmoo @kristinashu @mavis I rewrote this PR in a way that _should_ have preserved the look, but under the hood uses very different code to effect the index page, so basically the entire PR needs a rereview =)

(I've also changed the test URL to https://foundation-mofostaging-pr-3427.herokuapp.com/en/blog/ to better reflect its intended place)

kristinashu commented:
Spacing looks good!

## Grey bar
I'm noticing a grey bar just below the main nav that appears at medium and narrow widths. @mmmavis I feel like we've seen this before? Something about reminisce of the mobile drop down sub nav?

Scree cap from PR:
<img width="693" alt="image" src="https://user-images.githubusercontent.com/1068783/61835791-25f63100-ae32-11e9-8785-6248e28bcf0d.png">

## Line length for intro text
At wide widths the intro text should only cross 8 of the 12 bootstrap columns, instead of all 12 like in the PR (this is the keep the line length short and easier to read). But at medium and narrow widths it should continue to fill all the columns (since line length is shorter). 
<img width="797" alt="image" src="https://user-images.githubusercontent.com/1068783/61835993-0d3a4b00-ae33-11e9-8bdc-7001f60eb809.png">

Pomax commented:
Hm, the page template is based on the primary page, minus the "free content" parts, do primary pages also show this incorrect line sizes at those breakpoints in the review app? If so, we might have a site-wide regression for these templates.

Pomax commented:
Aha! Probably just this template's fault: https://github.com/mozilla/foundation.mozilla.org/pull/3427/files#diff-aa22ca63badaf6d5490c9347c4e8a92eR13 uses only `col-12` instead of the additional breakpoint size instructions.

Pomax commented:
@kristinashu fixed the line widths, it should now be 8 at medium and above, and 12 at smaller than that.

Pomax commented:
The grey bar looks like the minisite nav without any content (because there is none) so I'll see if I can remove that nav from this template entirely.

Pomax commented:
it was - removed the minisite-specific bits because this is not a minisite and what were they doing in the template in the first place here =P

Pomax commented:
@kristinashu @mavis the Percy failure looks like a genuine concern: the "participation guidelines" gets spread out over two columns...

kristinashu commented:
Why is this PR changing the footer?

Pomax commented:
Your guess is as good as mine: it's only showing changes in the footer for two (out of eleven?) pages, both entirely unrelated to this, and if I check those pages on the review app, things look perfectly fine, so.... it could just be Percy being wrong? =/

kristinashu commented:
Hmm weird, should we push to staging and then check those pages there?

Pomax commented:
That feels reasonable, although @mmmavis and @alanmoo will still need to R+ before we land.

mmmavis commented on network-api/networkapi/wagtailpages/templates/wagtailpages/fragments/generic-card.html:
```diff
@@ -0,0 +1,16 @@
+
+<div class="child-card col-6 col-md-4 mb-5">
+  <div>
+    <a href="{{ page.url }}"">
```
extra `"`

mmmavis replied:
Does it need quotation marks at all though?

Pomax replied:
it does, this isn't React JSX, Django templating does not automagically add the necessary quotes because it has no idea "where" the macro is (it can't tell the difference between `<a {{ href }}>` and `<a href="{{ href }}">`, which React can).

Pomax replied:
```suggestion
    <a href="{{ page.url }}">
```

alanmoo commented:
> Two questions:
> 
>     * In order for a page to show up on an index page, that page has to be child of the index page. This means child page urls will become `/index-page/child-page`. Will we running into issues with URL if we are planning to create index page of existing pages on production (e.g., their URL will change and we'll have to deal with redirects???)?

We shouldn't; We've had mini-site namespace pages in place for a while, so this will be replacing some of those, making previously dead URLs into real ones.

> 
>     * Will there be cases where we want a page to belong to more than one index page?

I think that's beyond the scope of intent for this page.

Pomax commented on network-api/networkapi/wagtailpages/templates/wagtailpages/fragments/generic-card.html:
```diff
@@ -0,0 +1,16 @@
+  <a class="h4-heading d-inline-block my-2" href={{ page.url }}>{{ page.title }}</a>
+  <p class="body-small mb-0">
+    {% block byline %}
+      Published on {{ page.first_published_at|date:"F j, Y" }}
```
Yeah, you're right.
